Jibri module

cleeyv opened the pull request:
###### Motivation for this change

I am working on [packaging jitsi](https://github.com/ngi-nix/ngi/issues/197) and all of its sub-components (including jibri) as part of the Summer of Nix 2021 program. There is also an open request for packaging jibri issue #132623 for which a package has been merged but not a module. This PR should be sufficient to close that issue. As I specified in the relevant commit messages, most of the code in this PR was written by @puckipedia for [nixcon-video-infra](https://github.com/nixcon/nixcon-video-infra) 2020. We have not been in communication yet about this PR but if she's reading this, any comments or suggestions are welcome!

###### Things done

This module as been successfully deployed and tested on a jitsi-meet instance with default configs, recording one hour long conference with four participants as well as multiple smaller recordings. The instance ran on a VM with 8GB of memory, with maximum memory usage during the call of around 4.5GB. Testing has been done with both the default config from `services.jitsi-meet.jibri.enable`, as well a more manual reimplementation of the same config directly using the options provided `services.jibri`. Both of tests required that jibri run on the same host that is also running jitsi-meet. I'm making this PR now because I think this is a reasonable default for nixpkgs, however the [jibri docs](https://github.com/jitsi/jibri#what-is-jibri) say that it is "intended to be run on a separate machine (or a VM)". It seems that this deployment of jitsi-meet and jibri to different hosts would require a refactoring of both the jibri module and the jibri-related elements of the jitsi-meet module (especially related to hostnames and passwords) which doesn't seem necessary for this PR to be merged. 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cleeyv committed:
nixos/jitsi-meet: add jibri.enable

This option enables a jibri service on the same host that is running
jitsi-meet. It was written, along with the jibri module, by @puckipedia
for nixcon-video-infra 2020.
Co-authored-by: Puck Meerburg <puck@puck.moe>

cleeyv committed:
jibri: add section to xorg conf file, for module

This change is adapted from a different jibri package, written by
@puckipedia for nixcon-video-infra 2020 along with a module for jibri.
Co-authored-by: Puck Meerburg <puck@puck.moe>

cleeyv committed:
nixos/jibri: init at 8.0-93-g51fe7a2

This module was written by @puckipedia for nixcon-video-infra 2020.
Minor changes made by @cleeyv for compat with existing jibri package.
Co-authored-by: Puck Meerburg <puck@puck.moe>

puckipedia commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+
+    recording = {
+      recordings-directory = "/tmp/recordings";
+      finalize-script = "/path/to/finalize"; # TODO(puck): replace with actual noop default
```
You might want to fix this no be a proper noop script, or possibly remove the TODO? jibri's not clear on the status of `finalize-script`

cleeyv replied:
Thanks for your feedback! I'm working on an [option](https://github.com/cleeyv/nixpkgs/commit/b44c85ddff8a57890a9c63baab09398b03921adb) for the module to allow for configuring a finalize script, using the [example script](https://github.com/jitsi/jibri/blob/master/resources/finalize_recording.sh) from the jibri repo as a default. While trying to test this, I discovered a bug in this module's jibri instance that I've been stuck trying to figure out: https://github.com/jitsi/jibri/issues/441 Did you notice anything like this when using jibri for nixcon last year? I've considered trying to run an instance from your nixpkgs fork in nixcon-video-infra to compare to what I'm seeing in the updated version, but I haven't tried it yet.

puckipedia replied:
I don't believe we had this issue; I suspect there's something else obvious that's missing -- jibri's basically being `^C`'d by another process, and that's all your logs indicate (both error code 130 and the SIGINT point to that)-- try checking the full journal at the moment of finalizing; worst case cross-reference the `si_pid` with an output of `ps` :p

puckipedia commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+
+
+    # Configure Chromium to not show the "Chrome is being controlled by automatic test software" message.
+    environment.etc."chromium/policies/managed/managed_policies.json".text = builtins.toJSON { CommandLineFlagSecurityWarningsEnabled = false; };
```
Annoyingly, this is a global setting, so this should possibly be an explicit flag/warning during rebuild, since it's a bit of a security issue(?) otherwise.

puckipedia commented on nixos/modules/services/web-apps/jitsi-meet.nix:
```diff
@@ -129,6 +133,16 @@ in
+      type = bool;
+      default = false;
+      description = ''
+        Whether to enable a Jibri instance and configure it to connect to Prosody.
```
Probably add a warning here to clarify that this is not supported procedure? :p

puckipedia commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+
+      script = (concatStrings (mapAttrsToList
+        (name: env: ''
+          export ${toVarName "${name}_control"}=$(cat ${env.control.login.passwordFile})
```
(is this space-safe?)

puckipedia commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+  toVarName = s: "XMPP_PASSWORD_" + stringAsChars (c: if builtins.match "[A-Za-z0-9]" c != null then c else "_") s;
+
+  defaultJibriConfig = {
+    id = "";
```
https://github.com/jitsi/jibri/blob/9871ba137cb759d22d31e5e1f7cfade8e8b4ca65/src/main/resources/reference.conf#L2-L4 suggests this will eventually(TM) change to require an ID at some point.

puckipedia reviewed (commented):
Some small nits! Happy to see someone try and upstream this :)

SuperSandro2000 commented on pkgs/servers/jibri/default.nix:
```diff
@@ -19,6 +30,8 @@ stdenv.mkDerivation rec {
     mv etc/jitsi/jibri/* $out/etc/jitsi/jibri/
     mv opt/jitsi/jibri/* $out/opt/jitsi/jibri/
 
+    cat '${xorgModulePaths}' >> $out/etc/jitsi/jibri/xorg-video-dummy.conf
```
```suggestion
    cp ${xorgModulePaths} $out/etc/jitsi/jibri/xorg-video-dummy.conf
```

SuperSandro2000 commented on nixos/modules/services/web-apps/jitsi-meet.nix:
```diff
@@ -256,6 +278,8 @@ in
         ${config.services.prosody.package}/bin/prosodyctl register focus auth.${cfg.hostName} "$(cat /var/lib/jitsi-meet/jicofo-user-secret)"
         ${config.services.prosody.package}/bin/prosodyctl register jvb auth.${cfg.hostName} "$(cat ${videobridgeSecret})"
         ${config.services.prosody.package}/bin/prosodyctl mod_roster_command subscribe focus.${cfg.hostName} focus@auth.${cfg.hostName}
+        ${config.services.prosody.package}/bin/prosodyctl register jibri auth.${cfg.hostName} "$(cat /var/lib/jitsi-meet/jibri-auth-secret)"
+        ${config.services.prosody.package}/bin/prosodyctl register recorder recorder.${cfg.hostName} "$(cat /var/lib/jitsi-meet/jibri-recorder-secret)"
```
Why not add it to the path?

SuperSandro2000 commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+      };
+    };
+
+    systemd.tmpfiles.rules =
+      [
+        "d /var/log/jitsi/jibri 755 jibri"
+      ];
+
+
```
```suggestion
    systemd.tmpfiles.rules = [
      "d /var/log/jitsi/jibri 755 jibri"
    ];
```

aanderse replied:
Please specify the group as well.

SuperSandro2000 commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      path = [ pkgs.chromedriver pkgs.chromium pkgs.ffmpeg-full ];
```
```suggestion
      path = with pkgs; [ chromedriver chromium ffmpeg-full ];
```

aanderse commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+{
+  options.services.jibri = with types; {
+    enable = mkEnableOption "Jitsi BRoadcasting Infrastructure";
+    config = mkOption {
```
We generally call this `settings`.

aanderse commented on nixos/modules/services/networking/jibri/default.nix:
```diff
@@ -0,0 +1,334 @@
+        config =
+          let
+            nick = mkDefault (builtins.replaceStrings [ "." ] [ "-" ] (
+              config.networking.hostName + optionalString (config.networking.domain != null) ".${config.networking.domain}"
```
We have `config. networking.fqdn` if that suits you.

aanderse reviewed (commented):
I left a few small points but unfortunately don't have the capacity to fully review this at this time.

cleeyv committed:
nixos/jibri: add finalize script option

cleeyv committed:
jibri: downgrade to jre8_headless from default v16

Fix for bug with jibri closing prematurely when recording is stopped:
https://github.com/jitsi/jibri/issues/441

cleeyv committed:
nixos/jibri: fix & docs for enable not via meet

cleeyv committed:
nixos/jibri: add nixos test

cleeyv committed:
nixos/doc: add jibri to release notes

cleeyv committed:
maintainers/jitsi: add cleeyv to jitsi team
